lib/recursive-methods.rb

```diff
@@ -1,49 +1,122 @@
+# frozen_string_literal: true
+
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# *****************************
+# Write a method factorial that accepts an integer parameter n and that uses recursion to compute and return the value of n factorial (also known as n!).
+
+# Time complexity: O(n) where n is n
+# Space complexity: O(1)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+
+  # e.g. fact(4) = 4 * 3 * 2 * 1 = 24
+  return 1 if n == 0
+
+  raise ArgumentError, "value must be 0 or greater" if n < 0
+
+  if n == 0
+    return n
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# *****************************
+# Write a method reverse that accepts a string as a parameter and then returns the reverse of the string by reversing all letters and all words in the string.
+
+# e.g. reverse("hello world") will return "dlrow olleh"
+
+# Time complexity: O(n) (I think? I'm not sure on this.)
```

Comment:
Because `s[0..-2]` creates a new string and you have n recursive calls this will be O(n^2) for both time and space.